### ISSUE TITLE
$product->getFinalPrice() ignores Catalog Price Rules in category view

### ISSUE THREAD

**Issue by infortis** [magento/magento2#13476]
### Preconditions

Magento 2.2.2.

#### 1
- Go to `Catalog > Products`. 
- Create **two** new products.
- For the first product set `SKU` to **TEST_01**.
- For the second product set `SKU` to **TEST_02**.
- For both products set `Price` $100 and `Special Price` $90 (with valid date). 
- Add both products to the same category.

#### 2
- Then, go to `Marketing > Catalog Price Rule`. 
- Create a new rule to apply a discount to selected product.
- Set `Actions > Apply` to `Apply as a percentage of original`.
- Enter `20` in the `Actions > Discount Amount`.
- Now we need to apply that rule only for the second product. In the `Conditions` of the rule, add the following condition (where **TEST_02** is the SKU of the second product):
```
If ALL of these conditions are TRUE:
SKU is TEST_02
```
- Save and apply the new rule.
- Reindex.

#### 3
Open `\vendor\magento\module-catalog\view\frontend\templates\product\list.phtml` and add this code:

```
	<?php
		$regular = $_product->getPrice();
		$final = $_product->getFinalPrice();
		echo "regular=" . $regular . ", ";
		echo "final=" . $final;
	?>
```

 somewhere inside the `foreach` loop, for example, before this line:

	<strong class="product name product-item-name">

#### 4
Flush the cache. Go to the frontend and open the category to which you've added the two newly created products.

### Expected result

Above the name of the first product we should see:
> regular=100.0000, final=90.0000

Above the name of the second product we should see:
> regular=100.0000, final=80.0000

The `final` price of the second product is **80.0000** because it is decreased by **20%** by the catalog price rule.

### Actual result

Above the name of the first product we see:
> regular=100.0000, final=90.0000

Above the name of the second product we see:
> regular=100.0000, final=90.0000

The `final` price of the second product is **90.0000** (which is the `Special Price` we've set for that product). So our new catalog price rule was completely ignored.

### The same code works correctly in product view

But if we add the same code to one of the template files of the product view, e.g. to `\vendor\magento\module-catalog\view\frontend\templates\product\view\addto.phtml`, then on the product page of the second product we will see:

> regular=100.0000, final=80.0000

So the same code returns **incorrect** final price **in category view**, but it returns correct final price in product view.

**Comment by AirmanAJK** [comment]
Just double-checking, but had the catalog rule been reindexed yet? If your indexing is done on a schedule, it could take however long your cron jobs are configured to run to see changes (recommended every minute though).

Try doing a full reindex on the Magento CLI to make sure and the flush all caches:
`indexer:reindex`

**Comment by AirmanAJK** [comment]
Product pages don't use at least several of the index tables and instead use the regular tables to gather data (catalog_product_flat for sure). I don't know if this includes the catalogrule_rule or catalogrule_product indexers, but it's likely an explanation for the behavior you are seeing.

**Comment by AirmanAJK** [comment]
Some last thoughts:
Look at the indexer table directly in the database and see what it has for your products (catalogrule_product_price).

Look at your cron_schedule table and make sure your reindex operations are running successfully. Error's here will be silent beyond this table's records since cron does it all in the background. I've had my scheduled indexing fail and didn't realize it for a while.

**Comment by magento-engcom-team** [comment]
@infortis, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by sedax90** [comment]
I can confirn the issue... some news for a patch?

**Comment by mokadev** [comment]
Any idea to fix this issue? Still present on 2.2.3

**Comment by mrodespin** [comment]
I have been testing this problem and i don't think is an issue because Magento show the correct final price "80.000" (following your steps)

The problem is that you are inserting a piece of code in the incorrect area.

Category product list price is rendered in "**view/base/templates/product/price/final_price.phtml**" template and the logic for category product listing is a little bit different than the product page.

In product page the code that calc the final prices with catalog rules is with the following function:

```
/**
     * Retrieve product final price
     *
     * @param float|null $qty
     * @param Product $product
     * @return float
     */
    public function getFinalPrice($qty, $product)
    {
        if ($qty === null && $product->getCalculatedFinalPrice() !== null) {
            return $product->getCalculatedFinalPrice();
        }

        $finalPrice = $this->getBasePrice($product, $qty);
        $product->setFinalPrice($finalPrice);

        **$this->_eventManager->dispatch('catalog_product_get_final_price', ['product' => $product, 'qty' => $qty]);

        $finalPrice = $product->getData('final_price');**
        $finalPrice = $this->_applyOptionsPrice($product, $qty, $finalPrice);
        $finalPrice = max(0, $finalPrice);
        $product->setFinalPrice($finalPrice);

        return $finalPrice;
}
```

That dispatch event triggers this observer:

```
<event name="catalog_product_get_final_price">
    <observer name="catalogrule" instance="Magento\CatalogRule\Observer\ProcessFrontFinalPriceObserver" />
</event>
```

With checks if the product is affected by any catalog rule in this file:

**Magento\CatalogRule\Observer\ProcessFrontFinalPriceObserver**

But for category product listing, this observer, is not triggered due cache implementations.

If you put this piece of code:

```
<?php
   $regular = $_product->getPrice();
   $final = $_product->getFinalPrice();
   echo "regular=" . $regular . ", ";
   echo "final=" . $final;
?>
```

Inside **/view/frontend/templates/product/list.phtml** you will not see the real final price, but if you put the next code in **/base/templates/product/price/final_price.phtml**, which is the template that renders the product price, you will see that the final price is correct.

```
<?php
    $regular = $block->getPrice()->getProduct()->getPrice();
    $final = $block->getPrice()->getProduct()->getFinalPrice();
     echo "regular=" . $regular . ", ";
     echo "final=" . $final;
?>
```

If you want to move the final price, because a frontend requirement, I recommend you use the category product blocks that renders  /base/templates/product/price/final_price.phtml or modify this code:

```
/**
     * Get product final price
     *
     * @param float $qty
     * @return float
     */
    public function getFinalPrice($qty = null)
    {
        if ($this->_getData('final_price') === null) {
            $this->setFinalPrice($this->getPriceModel()->getFinalPrice($qty, $this));
        }
        return $this->_getData('final_price');
    }
```

to

```
/**
     * Get product final price
     *
     * @param float $qty
     * @return float
     */
    public function getFinalPrice($qty = null)
    {
        $this->setFinalPrice($this->getPriceModel()->getFinalPrice($qty, $this));
    }
```

inside this file: app/code/Magento/Catalog/Model/Product.php

But i don't think that Magento will accept this.

#distributed-cd

**Comment by rksh1997** [comment]
I am still having the issue on Magento 2.3.0.

    $price = $product->getPrice(); // 45.00
    $special = $product->getSpecialPrice(); // NULL
    $discounted = $catalogRule->calcProductPriceRule($product,$product->getPrice()); // NULL
    $final = $product->getFinalPrice(); // 45.00

Looking at the database directly, the rule is applied and the price is 36.00 in `catalogrule_product_price`.

@mikewhitby

**Comment by AlexWorking** [comment]
Hello @infortis. We retested Your issue on Your reported version 2.2.2 and on the latest 2.2-develop and 2.3-develop versions. The described behaviour is realy reproduced on Your version, but on those latest develops it isn't reproduced  anymore. Its highly likely that the issue has been fixed, (if it was an issue at all -https://github.com/magento/magento2/issues/13476#issuecomment-375899773). Therefore currently we are closing this issue. Fell free to provide addition information/steps how it can be reproduced on lates 2.3 and reopen it if deem necessary.
